clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

```diff
@@ -0,0 +1,453 @@
+  // functions except the LLVM op will never set errno while the math library
+  // might. Also, math builtins have the same semantics as their math library
+  // twins. Thus, we can transform math library and builtin calls to their
+  // LLVM counterparts if the call is marked 'const' (known to never set errno).
+  // In case FP exceptions are enabled, the experimental versions of the
+  // intrinsics model those.
+  [[maybe_unused]] bool constAlways =
+      getContext().BuiltinInfo.isConst(builtinID);
+
+  // There's a special case with the fma builtins where they are always const
+  // if the target environment is GNU or the target is OS is Windows and we're
+  // targeting the MSVCRT.dll environment.
+  // FIXME: This list can be become outdated. Need to find a way to get it some
+  // other way.
+  switch (builtinID) {
+  case Builtin::BI__builtin_fma:
+  case Builtin::BI__builtin_fmaf:
+  case Builtin::BI__builtin_fmal:
+  case Builtin::BIfma:
+  case Builtin::BIfmaf:
+  case Builtin::BIfmal:
+    cgm.errorNYI("FMA builtins");
+    break;
+  }
+
+  bool constWithoutErrnoAndExceptions =
+      getContext().BuiltinInfo.isConstWithoutErrnoAndExceptions(builtinID);
+  bool constWithoutExceptions =
+      getContext().BuiltinInfo.isConstWithoutExceptions(builtinID);
+
+  // ConstAttr is enabled in fast-math mode. In fast-math mode, math-errno is
+  // disabled.
+  // Math intrinsics are generated only when math-errno is disabled. Any pragmas
+  // or attributes that affect math-errno should prevent or allow math
+  // intrincs to be generated. Intrinsics are generated:
+  //   1- In fast math mode, unless math-errno is overriden
+  //      via '#pragma float_control(precise, on)', or via an
+  //      'attribute__((optnone))'.
+  //   2- If math-errno was enabled on command line but overriden
+  //      to false via '#pragma float_control(precise, off))' and
+  //      'attribute__((optnone))' hasn't been used.
+  //   3- If we are compiling with optimization and errno has been disabled
+  //      via '#pragma float_control(precise, off)', and
+  //      'attribute__((optnone))' hasn't been used.
+
+  bool constWithoutErrnoOrExceptions =
+      constWithoutErrnoAndExceptions || constWithoutExceptions;
+  bool generateIntrinsics =
+      (constAlways && !optNone) ||
+      (!getLangOpts().MathErrno &&
+       !(errnoOverriden.has_value() && errnoOverriden.value()) && !optNone);
+  if (!generateIntrinsics) {
+    generateIntrinsics =
+        constWithoutErrnoOrExceptions && !constWithoutErrnoAndExceptions;
+    if (!generateIntrinsics)
+      generateIntrinsics =
+          constWithoutErrnoOrExceptions &&
+          (!getLangOpts().MathErrno &&
+           !(errnoOverriden.has_value() && errnoOverriden.value()) && !optNone);
+    if (!generateIntrinsics)
+      generateIntrinsics =
+          constWithoutErrnoOrExceptions && errnoOverridenToFalseWithOpt;
+  }
+
+  if (generateIntrinsics) {
```

Comment:
The code to compute the `generateIntrinsics` condition (from line 167 to here) was extremely tricky to get right, and I wouldn't be surprised if it changes as new pragmas are added. This really should be shared with classic codegen. Can you add a comment to that effect?